### ISSUE TITLE
Windows (x64) compiled executable (tesseract.exe) doesn't work in any other windows machine (x64)

### ISSUE THREAD

**Issue by jkang-eng** [tesseract-ocr/tesseract#2674]
### Environment

* **Tesseract Version**: tesseract 5.0.0-alpha-375-g179c8
* **Platform**: Windows Server 2008 R2, 64bit

### Current Behavior:
I built tesseract executable from source using following [this instruction](https://github.com/tesseract-ocr/tesseract/wiki/Compiling#develop-tesseract)

1. git clone https://github.com/tesseract-ocr/tesseract tesseract
2. Modify source code (this is to suit our application's purpose, certain that this is NOT the root cause of issue)
3. cd tesseract
4. cppan
5. mkdir build && cd build
6. cmake ..
7. cmake --build . --config Release

After step 6, I see tesseract.exe under "tesseract\build\bin\Release" directory, and executing it without any issue (working perfectly fine).

However, if once I copy entire **\Release** directory to another Windows with x64 platform (Windows Server 2008, 2016, etc..)  **It doesn't work anywhere else.**  (see screenshots below)

I have installed "Microsoft Visual C++ 2015-2019 Redistributable (x64)", **so dependency does not look like issue here.**

Screenshot 1 (**Installed runtime libraries**)
![2019-09-24 12_30_06-VirtualBoxVM](https://user-images.githubusercontent.com/55756666/65531894-e519a780-dec8-11e9-8b22-2ded63cc28c4.png)

Screenshot 2 (**Message box pop-up on different Windows machine**)
![2019-09-24 12_30_56-VirtualBoxVM](https://user-images.githubusercontent.com/55756666/65531907-e8149800-dec8-11e9-844e-952a80c23a45.png)

Screenshot 3 (**Windows Event Viewer**)
![2019-09-24 12_32_38-VirtualBoxVM](https://user-images.githubusercontent.com/55756666/65531941-f367c380-dec8-11e9-8d6a-c5a2e2226ca9.png)

Screenshot 4 (**Windows Event Viewer**) - FYI, disk is not on network, it's under C:
![2019-09-24 12_32_23-VirtualBoxVM](https://user-images.githubusercontent.com/55756666/65531947-f662b400-dec8-11e9-8a88-e9f3006c61f3.png)

Can anyone suggest me with a fix? is there anything I'm missing during compilation procedure?

**Comment by stweil** [comment]
CMake builds are currently optimized for the build host architecture. So if you build on a PC with AVX2 hardware, the resulting binaries won't work (= will crash) on a PC without AVX2.

**Comment by stweil** [comment]
Builds using `autoconf` will work.

**Comment by jkang-eng** [comment]
@stweil Thanks for your suggestions.
Regarding your recommendation ``CMakeLists.txt``, I'm compiling without following section in CMakeLists.txt and it seems to work on other Windows machine. 

When you mentioned build using ``autoconf``, can you instruct me how to do this?  I couldn't find ``autoconf`` usage under Windows page.

**Comment by zdenop** [comment]
With disabling optimization OCR process will be slow. 
Autotools macros are common for linux and using them on Windows is very difficult.
Other option is to use [cross-compiled windows build](https://github.com/UB-Mannheim/tesseract/wiki), but IMO it is not suitable for using tesseract API (library) on windows (VS Studio).

**Comment by jkang-eng** [comment]
> With disabling optimization OCR process will be slow.
> Autotools macros are common for linux and using them on Windows is very difficult.
> Other option is to use [cross-compiled windows build](https://github.com/UB-Mannheim/tesseract/wiki), but IMO it is not suitable for using tesseract API (library) on windows (VS Studio).

Do you have any guess what performance degradation would be without optimization ?  (I will need to test performance anyways but was just curious if you can make any rough guess)

Using cross-compiled windows build would be a problem for me, because I needed to changed the source code.

**Comment by zdenop** [comment]
With image from issue 263 I got this results (tesseract 5.0.0-alpha-447-g52cf, Intel Core i7-6600U 2,60GHz 2.8 GHz; 8 GB ram; Windows 64 bit) - 5 runs:

With autooptimization: 26.177767169865813
Without autooptimization: 50.47452997387488

**Comment by zdenop** [comment]
Now you can turn off cmake autooptimization with:
`cmake .. -DAUTO_OPTIMIZE=OFF`

**Comment by stweil** [comment]
@zdenop, did you specify `CMAKE_BUILD_TYPE` for your builds?

If not: please try latest Git master. It should be significantly faster.

**Comment by zdenop** [comment]
@stweil: I used cmake&ninja&clang and this combinations needs to specify `CMAKE_BUILD_TYPE`. So yes, I specify it (to Release).

**Comment by jkang-eng** [comment]
@zdenop Hi, thanks for testing the performance with optimization on/off.

> Update: I posted more details in #263

In your post, you mentioned
* tessdata_best
* tessdata_fast
* tessdata

Does this mean if I use eng.traineddata from each git, OCR engine will perform differently based on trained data file?  In other words, if I swap exsiting eng.traineddata with the one from "tessdata_fast", would it increase performance?

After reading some doc, it looks like eng.traineddata has effect when using LSTM engine.  I don't explicitly specify engine mode when executing.  is LSTM used by default?

**Comment by zdenop** [comment]
> if I swap existing eng.traineddata with the one from "tessdata_fast", would it increase performance?

I am not sure how do you measure performance, but `tessdata_fast` model is faster version of `tessdata_best`. It has some limitations (e.g. you can do training only from best model, and best model provide better OCR result - at least in some cases).

> is LSTM used by default?

```
>tesseract --help-oem
OCR Engine modes:
  0    Legacy engine only.
  1    Neural nets LSTM engine only.
  2    Legacy + LSTM engines.
  3    Default, based on what is available.
```
what is available => `tessdata_best` and `tessdata_fast` have only LSTM models (you will get error if you will use `--oem 0`: `Failed loading language 'eng' Tesseract couldn't load any languages! Could not initialize tesseract`), `tessdata` has also legacy model.

**Comment by jkang-eng** [comment]
@zdenop  Thanks, that makes sense.

Quick question regarding compilation (will create a new ticket if needed)

I'm using cmake to compile (shown below), which generate 64-bit executable.  Is there any way to build 32-bit env executable?
```
:: Compile Tesseract
cd build\tesseract_src
cppan

cd ..
cmake tesseract_src
cmake --build . --config Release
```

**Comment by zdenop** [comment]
1. IMO original topic is solved, so I will close this issue.
2. Issue tracker is for bug and not for support. Please use forum or better - read cppan doc. It is external project.

**Comment by jkang-eng** [comment]
Got it.  Thanks for the help.
